src/providers/recentmessages/RecentMessagesProvider.cpp

```diff
@@ -0,0 +1,218 @@
+        }
+
+        return messages;
+    }
+
+    void doMessageLoad(std::weak_ptr<Channel> channel)
+    {
+        auto ownedChannel = channel.lock();
+        if (!ownedChannel)
+        {
+            // Channel doesn't exist anymore so no need to load messages anyway
+            return;
+        }
+
+        if (!getSettings()->loadTwitchMessageHistoryOnConnect)
+        {
+            return;
+        }
+
+        QUrl url(Env::get().recentMessagesApiUrl.arg(ownedChannel->getName()));
+        QUrlQuery urlQuery(url);
+        if (!urlQuery.hasQueryItem("limit"))
+        {
+            urlQuery.addQueryItem(
+                "limit",
+                QString::number(getSettings()->twitchMessageHistoryLimit));
+        }
+        url.setQuery(urlQuery);
+
+        // We hold a reference to the channel anyway, might as well use it
+        NetworkRequest(url)
+            .onSuccess([ownedChannel](NetworkResult result) -> Outcome {
+                auto root = result.parseJson();
+                auto messages = parseRecentMessages(root);
+
+                auto &handler = IrcMessageHandler::instance();
+
+                std::vector<MessagePtr> allBuiltMessages;
+                allBuiltMessages.reserve(messages.size());
+
+                for (auto message : messages)
+                {
+                    if (message->tags().contains("rm-received-ts"))
+                    {
+                        QDate msgDate = QDateTime::fromMSecsSinceEpoch(
+                                            message->tags()
+                                                .value("rm-received-ts")
+                                                .toLongLong())
+                                            .date();
+                        if (msgDate != ownedChannel->lastDate_)
+                        {
+                            ownedChannel->lastDate_ = msgDate;
+                            auto msg = makeSystemMessage(
+                                QLocale().toString(msgDate,
+                                                   QLocale::LongFormat),
+                                QTime(0, 0));
+                            msg->flags.set(MessageFlag::RecentMessage);
+                            allBuiltMessages.emplace_back(msg);
+                        }
+                    }
+
+                    auto builtMessages = handler.parseMessageWithReply(
+                        ownedChannel.get(), message, allBuiltMessages);
+
+                    for (auto builtMessage : builtMessages)
+                    {
+                        builtMessage->flags.set(MessageFlag::RecentMessage);
+                        allBuiltMessages.emplace_back(builtMessage);
+                    }
+                }
+
+                postToThread([ownedChannel, root,
+                              messages =
+                                  std::move(allBuiltMessages)]() mutable {
+                    ownedChannel->addMessagesAtStart(messages);
```

Comment:
I don't really like having the loading implementation doing the adding of messages to the channel -- I think it's better to keep that logic separate in `TwitchChannel` and just have this class handle the loading and parsing of the messages.

Somewhat related to how I implemented the #3878, so not a big deal right now.

Reply:
Fair point but as you said, I would like to wait for #3878 to be merged in before making any major design changes